.changeset/slow-items-heal.md

```diff
@@ -0,0 +1,38 @@
+---
+"astro": minor
+---
+
+Adds experimental JSON Schema support for content collections of with `type: 'data'`.
+
+To enable this feature, add the experimental flag:
+
+```diff
+import { defineConfig } from 'astro/config';
+
+export default defineConfig({
+	experimental: {
++		contentCollectionJsonSchema: true
+	}
+});
+```
+
+To make sure VSCode picks up the schema file, you have multiple option. You can either reference the file in every file of your content collection:
+```diff
+{
++  "$schema": "../../../.astro/collections/test.schema.json",
+  "test": "test"
+}
+```
+
+Or you can set your VSCode settings to match them, read more in the [VSCode docs](https://code.visualstudio.com/docs/languages/json#_json-schemas-and-settings):
```

Comment:
``````suggestion
Then, add a `$schema` property to any data collection in `src/content.config.ts`:

```diff
// src/content/config.ts
import { defineCollection, z } from 'astro:content';

const test = defineCollection({
	type: 'data',
	schema: z.object({
+		"$schema": z.string().optional(),
		test: z.string()
	}),
});

export const collections = { test };
```

This experimental implementation requires you to manually reference the schema in each data entry file of the collection:

```diff
// src/content/test/entry.json
{
+  "$schema": "../../../.astro/collections/test.schema.json",
  "test": "test"
}
```

Alternatively, you can set this in your [VSCode `json.schemas` settings](https://code.visualstudio.com/docs/languages/json#_json-schemas-and-settings):
``````